Implement multiple span support for DocumentQuestionAnswering

ankrgyl opened the pull request:
# What does this PR do?

Extends DocumentQuestionAnswering pipeline to include support for multiple spans.

Fixes #18414


## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@NielsRogge @sgugger @Narsil

Commit:
Implement multiple span support

ankrgyl commented:
I still need to extend the tests to set the `max_seq_len` to be low (e.g. 10) and make sure that they work.

@NielsRogge another thing we could do in this change, or a separate one if y'all prefer, is to support multiple pages. The [DocQuery pipeline](https://github.com/impira/docquery/blob/main/src/docquery/ext/pipeline_document_question_answering.py#L227) already does this, because it's common to have multi-page documents and quite useful to be able to search across them. It's important to have this in the pipeline itself, so that you can properly handle the "no answer" score and top-k across pages. The main consideration is how we change the input shape of the pipeline to accommodate it, since we currently accept a (single) image and (optional) list of word_boxes. The options I can think of are:

- Image can also be a list of images, and then `word_boxes` can be a list of list of word boxes
- Combine image and `word_boxes` into one list of `(image, [(word, box)])` tuples
- Add a new optional argument, called pages, and use that instead of the image/word_box arguments if it's set.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19204). All of your documentation changes will be reflected on that endpoint.

sgugger commented on src/transformers/pipelines/document_question_answering.py:
```diff
@@ -305,44 +329,49 @@ def preprocess(self, input, lang=None, tesseract_config=""):
             # We put 0 on the tokens from the context and 1 everywhere else (question and special tokens)
             # This logic mirrors the logic in the question_answering pipeline
             p_mask = [[tok != 1 for tok in encoding.sequence_ids(span_id)] for span_id in range(num_spans)]
+            print(num_spans)
```
That's some debugging statement that should be removed, right?

ankrgyl replied:
Oops will remove! I was manually verifying that it worked :)

sgugger commented on src/transformers/pipelines/document_question_answering.py:
```diff
@@ -399,41 +433,38 @@ def postprocess_donut(self, model_outputs, **kwargs):
+        for output in model_outputs:
+            words = output["words"]
+
+            starts, ends, scores, min_null_score = select_starts_ends(
+                output["start_logits"],
+                output["end_logits"],
+                output["p_mask"],
+                output["attention_mask"].numpy() if output.get("attention_mask", None) is not None else None,
+                min_null_score,
+                top_k,
+                handle_impossible_answer,
+                max_answer_len,
+            )
```
Nit, using keyword arguments here would make the code easier to maintain.

ankrgyl replied:
I can change that! Good idea.

sgugger reviewed (approved):
I will let @NielsRogge dive deeper in this, but looking good on my side, thanks!

Commit:
Address comments

ankrgyl commented:
Absolutely. Apologies for the delay -- I will update the PR within the next few days.

Commit:
Add tests + fix bugs

ankrgyl commented:
@Narsil apologies again for the delay. I wanted to make sure I had some time to test it thoroughly. I added tests for the two flavors of LayoutLM models, but not for Donut, since there are no spans for Donut (i.e. there is no parameter or argument I can provide that hits two different code paths with the Chunk Pipeline, at least as far as I can tell).

I found several small bugs along the way and fixed them too.

Narsil reviewed (approved):
LGTM.

Thank you !

sgugger reviewed (approved):
Thnaks for working on this!

ydshieh commented:
Hi @ankrgyl Thank you for this PR. We need some help for CI failure regarding this new addition 🙏 

After this PR, we have 5 CI failures
```bash
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_chunk - AssertionError: Lists differ: [{'score': 0.9974, 'answer': '1110212019', 'start': 23, [69 chars] 16}] != [{'score': 0.9967, 'answer': '1102/2019', 'start': 22, '[67 chars] 15}]
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_donut - TypeError: forward() got an unexpected keyword argument 'page'
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_layoutlm_chunk - IndexError: index out of range in self
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_pt_LayoutLMConfig_LayoutLMForQuestionAnswering_LayoutLMTokenizerFast_nofeature_extractor - IndexError: tuple index out of range
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_pt_LayoutLMConfig_LayoutLMForQuestionAnswering_LayoutLMTokenizer_nofeature_extractor - IndexError: tuple index out of range
```
  - (For `test_large_model_pt_layoutlm_chunk`, I got different error from our CI where it is just output not equal expected value)
  - The 2 tests with `_chunk` at the end are new tests. The other 3 tests worked well in previous commits before this PR

Could you take a look, please 🙏 ?

Locally, I got one more test failure (not appear on CI), which also fails with previous commit (locally, not on CI). However, you can focus on the above 5 tests.
```
FAILED n tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_layoutlm - IndexError: index out of range in self
```

ankrgyl commented:
I'm so sorry about that! I will take a look ASAP.
